### ISSUE TITLE
Blurred font on HiDPI devices

### ISSUE THREAD

**Issue by reclaimingmytime** [loebi-ch/jquery-clock-timepicker#14]
On HiDPI devices, the font for part of the switcher looks blurred.

The following screenshot was taken on a 5.5-inch 1080p phone. You can see the font of the clock is blurred:
<img src="https://user-images.githubusercontent.com/35609464/36209053-f31ba0ec-119a-11e8-96e5-65175f705cea.jpg" alt="HiDPI" width="303" height="396">

This is what the font looks like in Firefox's responsive design view, where the clock looks sharper:
<img src="https://user-images.githubusercontent.com/35609464/36209298-a99146ec-119b-11e8-82ea-aae0f80d4622.png" alt="normal">

**Comment by dreuff** [comment]
Hello, add this to the end of the "INITIALIZE CLOCK CANVASES" section :

```
var dpr = window.devicePixelRatio || 1;
var hourCanvas = clockHourCanvas.get(0);
var minuteCanvas = clockMinuteCanvas.get(0);                                        
hourCanvas.width = canvasSize * dpr;
hourCanvas.height = canvasSize * dpr;
minuteCanvas.width = canvasSize * dpr;
minuteCanvas.height = canvasSize * dpr;                      
var hourCtx = hourCanvas.getContext('2d');
var minuteCtx = minuteCanvas.getContext('2d');
hourCtx.scale(dpr, dpr);
minuteCtx.scale(dpr,dpr);
```

**Comment by reclaimingmytime** [comment]
Thank you for the suggestion.

Unfortunately, this code did not resolve the issue on my OnePlus 5.

**Comment by loebi-ch** [comment]
@reclaimingmytime Have you found a solution for this issue?

**Comment by reclaimingmytime** [comment]
@loebi-ch I have not found a solution yet.

**Comment by Yauheni-Butski** [comment]
@reclaimingmytime , @loebi-ch , @dreuff 
I've found a solution for this issue.

Short answer:
Add the next code to the end of the "adjustMobilePopupDimensionAndPosition()" function. (section "ADJUST POPUP DIMENSION AND POSITION (FOR MOBILE PHONES)")

```
var dpr = window.devicePixelRatio || 1;
var hourCanvas = clockHourCanvas.get(0);
var minuteCanvas = clockMinuteCanvas.get(0);
hourCanvas.width = canvasSize * dpr;
hourCanvas.height = canvasSize * dpr;
minuteCanvas.width = canvasSize * dpr;
minuteCanvas.height = canvasSize * dpr;
var hourCtx = hourCanvas.getContext('2d');
var minuteCtx = minuteCanvas.getContext('2d');
hourCtx.scale(dpr, dpr);
minuteCtx.scale(dpr, dpr);

clockHourCanvas.css('width', canvasSize);
clockHourCanvas.css('height', canvasSize);
clockMinuteCanvas.css('width', canvasSize);
clockMinuteCanvas.css('height', canvasSize);
```

Long answer:
Thanks to @dreuff for his code, because it gave me a huge benefit in investigation. And also, as you can see, the different between his code and my is just in 4 last lines.

1. @dreuff 's code didn't work for me in "INITIALIZE CLOCK CANVASES" section, because function 'showTimePicker()' always calls 'adjustMobilePopupDimensionAndPosition()' which recalculate canvas' size.
2. @dreuff 's code worked perfectly in 'adjustMobilePopupDimensionAndPosition()' but it made canvas's size bigger (if devicePixelRatio is 2).
3. Based on this link https://www.html5rocks.com/en/tutorials/canvas/hidpi/ I've added the last 4 lines. Quote: "The de facto solution to this has been to create images scaled up by the devicePixelRatio and then use CSS **to scale it down by the same amount**, and the same is true for canvas!".
4. So, when initial canvas's size is setting, for example to '500px' when dpr=2 in style attribute, we set  '250px' size via CSS.

I've tested it on Android and iOS devices and it looks like it works.

If you found any issues with this solution - please, feel free to leave a comment.

**Comment by reclaimingmytime** [comment]
@Yauheni-Butski This seems to resolve the issue. Adding the code at the correct position makes the font clear again.

The only comment I have is that some code from your solution is already present, although in a slightly different form.

In particular, the following lines:

```javascript
clockHourCanvas.css('width', canvasSize);
clockHourCanvas.css('height', canvasSize);
clockMinuteCanvas.css('width', canvasSize);
clockMinuteCanvas.css('height', canvasSize);
```

...would remove the need for the existing code just at the end of the `adjustMobilePopupDimensionAndPosition` method:
```javascript
clockHourCanvas.attr('width', canvasSize);
clockHourCanvas.attr('height', canvasSize);
clockMinuteCanvas.attr('width', canvasSize);
clockMinuteCanvas.attr('height', canvasSize);
```
If the existing code is kept in place, then the `height` and `width` attributes to `clockHourCanvas` and `clockMinuteCanvas` are set twice, once by the existing code and once by the fix.

This is only a performance optimization, however, and does not affect the result of the fix.

**Comment by loebi-ch** [comment]
@reclaimingmytime I remarked the same. But since I cannot reproduce the issue on my mobile, can you please let me know what's the best optimized fixed for this issue so I can implement it in the new version correctly. Thank you.

**Comment by Yauheni-Butski** [comment]
@reclaimingmytime yes, I noticed that some lines repeat existed in above and we re-set width the second time. Just wanted to provide quick, whole fix without some refactoring.
 So yes, you are right that some lines can be changed.

**Comment by reclaimingmytime** [comment]
@loebi-ch The only optimization I could get to work was the one mentioned above. So, keep the `.css` from the fix and comment/remove the `.attr([...], canvasSize)` just above.

---
**Edit:** Below is a suggestion for a possible performance improvement, which unfortunately I did not get to work - the font becomes blurred again. I could not find any explanation for this yet, so the best course of action would be to go without this for now.

After taking a look at the rest of the code, I noticed the width and height attributes are already set in the _INITIALIZE CLOCK CANVASES._ section ([lines 361 and 362](https://github.com/loebi-ch/jquery-clock-timepicker/blob/master/jquery-clock-timepicker.js#L338-L339)). This is done using the `attr` function.

In order to keep changing the width and height more consistent with the rest of the code, the `.attr` function should be kept instead of using the `.css` function for the fix. Otherwise, the clock would have a `width` and `height` attribute in addition to an inline style which sets a different width and height, i.e. look somewhat like this:
```html
<canvas class="clock-timepicker-hour-canvas" width="500" height="500" style="[...]width: 250px; height: 250px;"></canvas>
```

So, the optimized code would look like this (the change being using `attr` instead of `css`): _[edit: Note again this does not work]_
```javascript
var dpr = window.devicePixelRatio || 1;
var hourCanvas = clockHourCanvas.get(0);
var minuteCanvas = clockMinuteCanvas.get(0);
hourCanvas.width = canvasSize * dpr;
hourCanvas.height = canvasSize * dpr;
minuteCanvas.width = canvasSize * dpr;
minuteCanvas.height = canvasSize * dpr;
var hourCtx = hourCanvas.getContext('2d');
var minuteCtx = minuteCanvas.getContext('2d');
hourCtx.scale(dpr, dpr);
minuteCtx.scale(dpr, dpr);

clockHourCanvas.attr('width', canvasSize);
clockHourCanvas.attr('height', canvasSize);
clockMinuteCanvas.attr('width', canvasSize);
clockMinuteCanvas.attr('height', canvasSize);
```

**Comment by loebi-ch** [comment]
@Yauheni-Butski Can you please have a quick check of version 2.3.3 at [https://plugins.slyweb.ch/jquery-clock-timepicker](https://plugins.slyweb.ch/jquery-clock-timepicker) if everything works like expected and give me a GO for the release. Thank you!

**Comment by Yauheni-Butski** [comment]
@loebi-ch I think I will be able to check it only on Monday. So if anybody else can do it - feel free! :)

**Comment by reclaimingmytime** [comment]
@loebi-ch Thank you for the new version.

I noticed the new version does not include the fix from pull request #13. Is this by mistake?

Regarding the fix, I could get it to work with my phone after changing the last few lines in the `adjustMobilePopupDimensionAndPosition` function. Specifically, I replaced the "attr" with the "css" function, e.g. so
```javascript
clockMinuteCanvas.attr('width', canvasSize)
```
becomes
```javascript
clockMinuteCanvas.css('width', canvasSize)
```
After making this change with all four lines, the fix worked for me.

With the current fix at https://plugins.slyweb.ch/jquery-clock-timepicker and the modifications mentioned above, the font is clear. The additional changes I had to make could be due to the way my browser prioritises styles, so it is best if someone else verifies this.

**Comment by loebi-ch** [comment]
@reclaimingmytime Thank you for checking and reminding me of the missing pull request #13.

Now it should be okay:
[https://plugins.slyweb.ch/jquery-clock-timepicker/](https://plugins.slyweb.ch/jquery-clock-timepicker/)

Last check :-)

**Comment by reclaimingmytime** [comment]
Now it works for me, thanks.

**Comment by loebi-ch** [comment]
Resolved in version 2.3.3
